### ISSUE TITLE
Term reordering follow up from #15985

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#16090]
I just found this:
```julia
In [14]: integrate(2 * sqrt(-x**2 + 1)*(4*sqrt(2)*x**2 - sqrt(3)), (x, -1, +1))                                                   
Out[14]: -√3⋅π + √2⋅π
```
Haven't looked into the details yet but do you have any idea why this happens @nickovs:
```julia
In [18]: sqrt(2)-1                                                                                                                
Out[18]: -1 + √2
```

_Originally posted by @oscarbenjamin in https://github.com/sympy/sympy/pull/15985#issuecomment-467681751_

**Comment by oscarbenjamin** [comment]
I had thought that those two examples would show up with as term1-term2 rather than -term1+term2 after #15985. Opening this issue to record examples for followup from there as [suggested](https://github.com/sympy/sympy/pull/15985#issuecomment-467687615).

**Comment by nickovs** [comment]
It seems like the most general way to do this would be to spot when the top level is an addition of two terms where one of the terms would be preceded by a minus if printed on its own but the other would not, and in that case print the negative one on the right.

Before I go off and do this I would like to have a discussion of what are the cases where this model would *not* render the most natural representation. Are there any situations where, printed on its own, people would expect `-(foo) + (bar)` rather than `(bar) - (foo)`? We know that there are cases where a set of results (such as solutions to an equation) are returned and we might prefer `{-x² + π, x² + π}` over `{π - x², x² + π}`. Are there other situation where putting the negative part on the right looks worse?

**Comment by supreet11agrawal** [comment]
What is the problem if we are getting the results in this manner. I don't understand

**Comment by nickovs** [comment]
@supreet11agrawal The goal here is for the printout from Sympy to be as close as possible to what a mathematician would naturally write. In most cases a statistician would never write `-p + 1` when she meant `1 - p`, and many people see `-1 + √2` as clunky and unnatural compared to `√2 - 1`. The representation of the former has already been fixed by my (rather narrowly targeted) PR #15985 but there are many other cases that still look bad.

To a first order approximation, having a leading minus in front of one term and following that term with a plus before one other term looks unnatural. Fixing it for all of those cases is easy. The question that we need to resolve is: do we want to change this for *all* of those cases or are there enough cases where `-(foo) + (bar)` looks better than `(bar) - (foo)` that we need a more subtle analysis?

**Comment by asmeurer** [comment]
Is it doing this because an Add of numeric terms is sorted numerically?

**Comment by oscargus** [comment]
The only case I can think of it the type of symmetry that you @nickovs is mentioning. Then, I assume most people would actually prefer `{π - x², π + x²}` which I guess just shows that it is a very hard problem to get it correct all of the time ("put constant to the left if there is another expression somehow printed at the same time where that is better", which may be feasible in e.g. sets or bounds, but sometimes those two equations are separate..., still for three terms it becomes even harder).

Hence, I believe that what you outline as a solution is probably the preferred one. As you cleverly states "it's going to be hard to please all of the cases all of the time" and, hence, going for one where the obviously "ugly" case happens less often is preferred.

I'm, not sure that I have followed all of the discussion, but wouldn't it be possible to look at arbitrarily long additions, not just two terms, and reorder the two first terms if the first one would be printed with a minus if on its own? Maybe one should introduce a helper function to determine if an expression in its printed form is starting with a minus, the simplest form being `str(expr)[0] == '-'`but probably one can determine it in more clever way. I can imagine a few more use cases for that (like moving a minus from inside an integral to outside, if that is not currently done).

**Comment by oscarbenjamin** [comment]
I've often wondered what it is but is this what `could_extract_minus_sign` is for?
```
could_extract_minus_sign() method of sympy.core.symbol.Symbol instance
    Return True if self is not in a canonical form with respect
    to its sign.
    
    For most expressions, e, there will be a difference in e and -e.
    When there is, True will be returned for one and False for the
    other; False will be returned if there is no difference.
    
    Examples
    ========
    
    >>> from sympy.abc import x, y
    >>> e = x - y
    >>> {i.could_extract_minus_sign() for i in (e, -e)}
    {False, True}
```

**Comment by nickovs** [comment]
@oscarbenjamin I'm not quite sure what `could_extract_minus_sign` is designed for but it does something slightly different to what we need. That said, looking at the implementation it seems like if I have a term `t` then `t.extract_multiplicatively(-1)` does exactly what we need.

At first glance It seems that replacing my original patch with:
```
        if order == None and self.is_Add:
            # Spot the special case of Add(Number, Mul(Number, expr)) with the
            # first number positive and thhe second number nagative
            add_args = Add.make_args(self)
            if len(add_args) == 2:
                key = lambda t: t.extract_multiplicatively(-1) is not None
                add_args = sorted(add_args, key=key)
                return add_args
```
does pretty much exactly what we need. It certainly fixes all of the expressions discussed here. I will run a bunch more tests on this. If it seems to do the trick then I will see which test cases it breaks and fix them up.

**Comment by nickovs** [comment]
@oscargus Regarding fixing up longer expressions, I think that displaying `x - x² + 1` would be less desirable than the current `-x² + x + 1` because having the polynomial terms out of order would be confusing and terms would get visually lost. In this case `1 + x - x²` might be better (i.e. reverse the whole thing) but when reversing additions with more than two terms the visual distance between two expressions that only differ in the sign of one term could become very large. Thus I suggest that for the time being we stick with reordering only when there are two terms.

**Comment by nickovs** [comment]
I've built a fix for this: https://github.com/nickovs/sympy/commit/6159952e8ccc4592b298531250d0f493f40278e6. It turns out that now that the code is a little more general it helps to perform the regular ordering first and then patch up the result afterwards if necessary. As a side effect this also reduces the total work and the code is smaller. 

I have not as yet fixed up all of the numerous doctests that are broken by this. Before I do so it would be helpful if interested parties could run the doctests, look at the changes and decide if they think that the results generally look better in the new form. In my opinion they do in pretty much all of the cases but it would be good to have confirmation before I slog through and change all of the doc strings, since that is a lot more work than writing the code!

**Comment by oscarbenjamin** [comment]
One way to demonstrate the doctest changes is by pushing a PR so we can see the failing doctest output on Travis.

**Comment by nickovs** [comment]
@oscarbenjamin I could do that. In general I don't like pushing half-done PRs that fail stacks of tests (and for which the test output ends with `DO *NOT* COMMIT!`). Also running a full battery of tests on travis takes sufficiently long as to actually cost someone a non-trivial amount of money for the CPU time and possibly warms the planet by a measurable amount :-)

Instead I have parked a copy of the output from just the doctests in this Gist: https://gist.github.com/nickovs/682a83ba6c02021ba600445cdee2f45d

**Comment by oscarbenjamin** [comment]
We've changed the way that Travis runs the tests now so it wouldn't use as much time as you think if it's based off recent master.

**Comment by oscarbenjamin** [comment]
What's happening [here](https://gist.github.com/nickovs/682a83ba6c02021ba600445cdee2f45d#file-doctest_output-L1083):
```
File "/Users/nicko/sympy/sympy/solvers/solveset.py", line 1431, in sympy.solvers.solveset._transolve
Failed example:
    pprint(tsolve(f, x), use_unicode=False)
Expected:
         /   3     \
         | --------|
         | log(2/9)|
    [-log\2         /]
Got:
        /    3    \ 
        | --------| 
        | log(9/2)| 
    [log\2        /]
```

**Comment by nickovs** [comment]
I've no idea. My code only re-orders the results of `as_ordered_terms()` so my best guess is that the solver is calling `as_ordered_terms()` without explicitly specifying `order=...` and making an assumption about the ordering that will be returned. I'd argue that that would be a bug in the solver not my code, but it's still clearly a bug.

**Comment by oscarbenjamin** [comment]
And [this one](https://gist.github.com/nickovs/682a83ba6c02021ba600445cdee2f45d#file-doctest_output-L3121):
```
File "/Users/nicko/sympy/doc/src/modules/rewriting.rst", line 95, in rewriting.rst
Failed example:
    pprint(cse((x-y)*(z-y) + sqrt((x-y)*(z-y)), optimizations='basic'),
        use_unicode=True)
Expected:
    ⎛                          ⎡  ____     ⎤⎞
    ⎝[(x₀, -(x - y)⋅(y - z))], ⎣╲╱ x₀  + x₀⎦⎠
Got:
    ⎛                          ⎡  ____     ⎤⎞
    ⎝[(x₀, -(y - x)⋅(z - y))], ⎣╲╱ x₀  + x₀⎦⎠
```

**Comment by oscarbenjamin** [comment]
I've been through all of those doctests. The vast majority of changes are improvements. There's a few cases where it looks a little less good but we can't win them all so I'm happy with this.

Need to follow up on those two cases though. It seems as if `as_ordered_terms()` is affecting the symbolic representation.

**Comment by nickovs** [comment]
That second one is correct; `-(y - x)⋅(z - y)` is equal to `-(x - y)⋅(y - z)`.

**Comment by oscarbenjamin** [comment]
I agree that it's correct but why has this change affected the result?

I thought this was just about ordering but there it's cancelled a minus sign between the two factors.

**Comment by nickovs** [comment]
I fear that the second one might have been correct only by luck.

Digging into the call history for calls to the new code from the `solveset.py` case, it seems that there calls to `signsimp()` that result in calls to expr.sort_key() without an explicit order and this calls the new code, and also calls from `parallel_poly_from_expr()` which end up calling `__str__()` and then applying a regex the text result (via `polys.polytuils.order_key()`). I will see if I can work out which is causing the issue.

It's worth noting that the very first fix that I produced placed all the changes in the printing code (in `printing.printer._as_ordered_terms()` but @asmeurer asked me to move it to the main `as_ordered_terms()` function in `expr.py` https://github.com/sympy/sympy/pull/15985#issuecomment-463866880. If moving it back fixes the problem then maybe we should revert to "Plan A", since my feeling is that this sort of prettier representation should be handled at the presentation layer. Of course when you have hacks in `parallel_poly_from_expr()` that rely the string representation to order terms it might be a problem even there.

**Comment by nickovs** [comment]
Moving the code back to `printing.printer._as_ordered_terms()` makes both of the problems identified above go away. The new set of output changes can be found here:
https://gist.github.com/nickovs/22e884760b02fa5040985b408a5dfd5e

**Comment by oscarbenjamin** [comment]
I think having this in `as_ordered_terms` is fine. As @asmeurer [said](https://github.com/sympy/sympy/pull/15985#issuecomment-463866880): it may even unearth some bugs.

Can you open a separate issue for cse? For this issue you can just change the doctest since it's no more or less correct than before.

**Comment by oscarbenjamin** [comment]
For tsolve both values are the same:
```julia
In [12]: x1 = -log(2**(3/log(S(2)/9)))                                                                                                                                                                      

In [13]: x2 = log(2**(3/log(S(9)/2)))                                                                                                                                                                       

In [14]: x1                                                                                                                                                                                                 
Out[14]: 
    ⎛    3    ⎞
    ⎜ ────────⎟
    ⎜ log(2/9)⎟
-log⎝2        ⎠

In [15]: x2                                                                                                                                                                                                 
Out[15]: 
   ⎛    3    ⎞
   ⎜ ────────⎟
   ⎜ log(9/2)⎟
log⎝2        ⎠

In [16]: x1.evalf()                                                                                                                                                                                         
Out[16]: 1.38253626185511

In [17]: x2.evalf()                                                                                                                                                                                         
Out[17]: 1.38253626185511
```

**Comment by nickovs** [comment]
Right. I hadn't spotted that it was `-log(... 2/9)` and `log(... 9/2)`, so these are the same value. So, it seems that the patch as proposed doesn't break any results, it only changes the way that the results are presented.

**Comment by smichr** [comment]
The run [here](https://travis-ci.org/smichr/sympy/jobs/541422079) shows what expressions look like when the two-term add is ordered, via printing, so the negated term comes second. This is the diff
```diff
diff --git a/sympy/printing/str.py b/sympy/printing/str.py
index ee560ca..cf1dcfd 100644
--- a/sympy/printing/str.py
+++ b/sympy/printing/str.py
@@ -51,6 +51,8 @@ def _print_Add(self, expr, order=None):

         PREC = precedence(expr)
         l = []
+        if len(terms) == 2 and str(terms[0])[0] == '-' and str(terms[1])[0] != '-':
+            terms.reverse()
         for term in terms:
             t = self._print(term)
             if t.startswith('-'):
```

**Comment by nickovs** [comment]
@smichr Thanks for that. The results looks much the same as my tests.

With regard to your code fix, it is certainly easy from a code point of view to do this by calling `str()` but that call can get very expensive when the argument is not a simple term. It's also worth noting that, since `str()` will get called again for both terms once the order is finalised, you'd be recursively calling `str()` twice for both terms, so worst case you can end up with a running time that is quadratic in the total number of terms, making an already slow call much slower. As such I'd like to avoid using `str()` to decide on the layout of `str()`.

Looking back at this thread, I left off while waiting for input as to if we wanted to proceed with the whole-sale reordering of all two-term additions where one term would be printed with a leading minus. The consensus seems to be yes but I was hoping for more discussion. In the absence of more people with opinions I will assume that most people like this and that I should proceed.

As before, the vast majority of the work here is not writing the code but fixing the doctests. I'm unlikely to get a chance to do this all very soon due to family vacation but I will see what I can do.

**Comment by oscarbenjamin** [comment]
I think it's agreed that this is a good idea.

**Comment by nickovs** [comment]
OK. I'll work on fixing up the doctests.

**Comment by oscarbenjamin** [comment]
This just came up on StackOverflow: https://stackoverflow.com/questions/57659019/re-arrange-sympy-output/57660615#57660615

**Comment by nickovs** [comment]
Sorry it's taken longer for me to get to this than I hoped. I changed jobs just around the time that it was agreed that the fix was a positive one. The remaining work is all around fixing up the doctests. I will see if I can find some time for that and submit a complete PR.

**Comment by oscarbenjamin** [comment]
There's no need to apologise. Thank you for working on this and I hope you do find the time.
